rfc/languages-in-w3s.md

```diff
@@ -0,0 +1,84 @@
+- The w3up repo is huge, and honestly, even if you know javascript, there's a lot to get up to speed on. We should have a guide to the code.
+
+### Go
+
+- Go is a highly performant garbage collected language like Java, but is much simpler than Java
+- Go has an extremely powerful concurrency model with green threads
+- These two components together make Go one of the most powerful languages for writing servers of any kind. The challenge of server programming is not super fast compute, but super high concurrency execution. Go excels at this, which matters cause our project has a lot of servers involved.
+- Go's ability to do binary distributions by default is a major plus for kind of project.
+- By and large, people get stuff done quickly in Go because Go is extremely simple. It's very fast to learn, easy to read, and it all looks the same. The only problem is its extremely boring. (Note from Hannah: When I first started writing Go I hated it. I thought someone took the joy out of programming. Now I love it cause I get my joy from what I'm building, not my programming language)
+- Lots and lots of Filecoin tooling and software is written in Go. In fact, W3S already has Go in it, we just don't maintain it -- it's called Spade
+- Garbage collection and green threads present challenges for using Go over an FFI, or for compiling go to WebAssembly, though TinyGo has improved this story by reducing the runtime size.
+
+### Rust
+
+Caveat from Hannah: I am not a highly experienced Rust dev
+
+- Rust is a better C/C++. Like definitely much better. This is huge cause no one wrote a better C/C++ for low level programming for 20 years.
+- However, by and large, you should probably not write programs in Rust unless you'd consider writing them in C/C++. There is minimal evidence one can be as productive in Rust as higher level languages.
+- In particular, you should probably not write web servers in Rust or any other asynchronous programs unless you have extra time on your hands which we don't.
+- BTW, if you haven't programmed in C/C++ recently, you've probably forgotten that long compile times are a thing on large projects in low level languages. 
+- Rust is essential for highly performant synchronous compute tasks. That's why people use it for smart contracts, hashing, and proofs.
+- Rust has the best WebAssembly story of any language.
+
+## Table Stakes For Considering A Language for W3S
+
+In addition to having a compelling engineering or business reason, we have some other things to consider before we adopt a language:
+
+- UCAN libraries and tooling especially for anything with a web interface
+- IPLD/IPFS/Libp2p libraries (we may not always need libp2p going forward)
```

Comment:
I feel like this write up overlooks one important reason why we have sticked to JS:

1. All the cloud services CF, AWS (pretty much anyone else) have a best support for JS and some support for other languages. And if we don't ant to think about those services even PLware  station  is also JS.
   - We may reduce dependencies on such things in the future, but flexibility to use them in tandem is important IMO.
   - Rust is probably gaining momentum in all of those also mostly because of WASM and because those cloud services use Rust themself (as they need to scale)

I would also consider IPVM even if Fission is going away I keep hearing IPVM will carry on and so will everywhere.computer Even if those projects discontinue there is a lot of excellent code that was written for those that could be a great foundation to build from.

Reply:
Fundamentally I need more signal on IPVM continuing onward before I'd commit anything we do to using it.

Reply:
100% agree on Javascript portability.

Reply:
> Use javascript when you really don't know where your code might run
... but you can be certain about the runtime environment (in browser, node-js version, etc.)

Use go to build executables when you know nothing about where it is running, only that it is not in a browser.